### ISSUE TITLE
Support for efficiently processing categorical distributions with varying dimensions

### ISSUE THREAD

**Issue by ArchieGertsman** [pytorch/pytorch#102938]
### 🚀 The feature, motivation and pitch

Say I have a list of probability distribution tensors of varying size. It would be nice if `Categorical` could accept a flattened tensor `probs_flat` containing all the distributions, along with a pointer `ptr` that indicates the starting index of each distribution. Standard functions can efficiently be computed, e.g. entropy can be computed as `-segment(probs_flat.log() * probs_flat, ptr)`, where `segment` is something like `segment_add_csr` from `torch_scatter` (I'm not sure if torch provides this function out of the box).

I'm not too sure of the best way to integrate this with the existing code. Just an idea!

### Alternatives

Let `probs_list` be a list of probability distribution tensors, and `xs` be a list of samples from their respective distributions. There are currently two options:
1. (evaluate for each)
```Python
log_probs = torch.stack([
    Categorical(probs=probs).log_prob(x)
    for probs, x in zip(probs_list, xs)
])
```
2. (pad, stack, evaluate)
```Python
max_size = max(probs.numel() for probs in probs_list)
probs_padded_list = []
for probs in probs_list:
    probs_padded = probs.new_zeros(max_size)
    probs_padded[:probs.numel()] = probs
    probs_padded_list += [probs_padded]
pi = Categorical(torch.stack(probs_padded_list))
log_probs = pi.log_prob(torch.stack(xs))
```

### Additional context

This can be useful in RL, where the number of possible actions varies between environment steps, as a less wasteful alternative to action masking.
